tensorflow/compiler/xla/service/gpu/amdgpu_compiler.cc

```diff
@@ -139,436 +135,101 @@ string GetROCDLDir(const HloModuleConfig& config) {
-  // In some cases, we have to place the result of an instruction in a temporary
-  // buffer. For instance, the buffer that holds an external parameter is
-  // assumed immutable at this point, and should not be reused for output
-  // (b/27180329). Therefore, in that case, we set the output to be a copy of
-  // the parameter.
-  HloPassPipeline pipeline("GPU-ir-emit-prepare");
+Status AMDGPUCompiler::OptimizeHloPostLayoutAssignment(
+    HloModule* hlo_module, se::StreamExecutor* stream_exec,
+    se::DeviceMemoryAllocator* device_allocator) {
+  HloPassPipeline pipeline("post-layout_assignment");
   pipeline.AddInvariantChecker<HloVerifier>(
       /*layout_sensitive=*/true,
       /*allow_mixed_precision=*/false,
       LayoutAssignment::InstructionCanChangeLayout);
 
-  // Copy insertion should be performed immediately before IR emission to avoid
-  // inserting unnecessary copies (later pass adds an instruction which
-  // materializes the value) or missing a necessary copy (later pass removes an
-  // instruction which materializes a value). DCE must be run immediately before
-  // (and sometime after) copy insertion, to avoid dead code from interfering
-  // with the rewrites.
-  pipeline.AddPass<HloDCE>();
-  pipeline.AddPass<FlattenCallGraph>();
-  pipeline.AddPass<GpuCopyInsertion>();
-  pipeline.AddPass<GpuSanitizeConstantNames>();
-
-  return pipeline.Run(hlo_module).status();
-}
-
-}  // namespace
-
-AMDGPUCompiler::AMDGPUCompiler()
-    : pointer_size_(llvm::DataLayout(amdgpu::kDataLayout)
-                        .getPointerSize(0 /* default address space */)),
-      platform_id_(se::rocm::kROCmPlatformId) {}
-
-AMDGPUCompiler::AMDGPUCompiler(se::Platform::Id platform_id)
-    : pointer_size_(llvm::DataLayout(amdgpu::kDataLayout)
-                        .getPointerSize(0 /* default address space */)),
-      platform_id_(platform_id) {}
-
-StatusOr<std::unique_ptr<HloModule>> AMDGPUCompiler::RunHloPasses(
-    std::unique_ptr<HloModule> module, se::StreamExecutor* stream_exec,
-    se::DeviceMemoryAllocator* device_allocator) {
-  // We dump the post-optimization HLO in RunBackend so no need to dump it here.
-  VLOG(3) << "*** HLO Before Optimization";
-  XLA_VLOG_LINES(3, module->ToString());
-
-  XLA_SCOPED_LOGGING_TIMER("AMDGPUCompiler::RunHloPasses");
-  tensorflow::profiler::TraceMe activity(
-      [&] { return absl::StrCat("HLO Transforms:", module->name()); },
-      tensorflow::profiler::TraceMeLevel::kInfo);
-  TF_RETURN_IF_ERROR(
-      OptimizeHloModule(module.get(), stream_exec, device_allocator, this));
+  // The LayoutAssignment pass may leave behind kCopy instructions which are
+  // duplicate or NOPs, so remove them with algebraic simplification and CSE.
+  AlgebraicSimplifierOptions options;
+  options.set_is_layout_sensitive(true);
+  pipeline.AddPass<HloPassFix<AlgebraicSimplifier>>(options);
+  pipeline.AddPass<MiopenConvAlgorithmPicker>(stream_exec, device_allocator);
```

Comment:
Do we want to keep the large chunk of comment on pick the algorithm before fusion?

Reply:
the comment is literally a copy/paste from cudnn counterpart and would surely be frowned at in upstreaming. I'll let you decide when you revise `MiopenConvAlgorithmPicker` :)